### ISSUE TITLE
Output probability matrix for each class

### ISSUE THREAD

**Issue by enchantress1016** [matterport/Mask_RCNN#583]
In my application, I need the output to be a matrix of probabilities for each class, such that each value in the matrix denotes the probability of this corresponding pixel to be in this class.
Could anybody elaborate on explaining how I could achieve this idea? Your help is really appreciated.

**Comment by zungam** [comment]
Im not an expert on this, but it sounds a bit tricky. Im not sure Mask RCNN would be much help you there because in how it works. As you see in the architecture below the classification comes before generating the pixel by pixel. 
![image](https://user-images.githubusercontent.com/11159388/40539498-36ac2c50-6015-11e8-9528-9312b19337d5.png)
I believe the mask part dosent "care" what kind of class it is, it just makes a mask for all types of objects.
This is my theory on how it works.

**Work around:**
You could train three different MASK-RCNN models but only feed one type of image for each model. Say for instance one duck-network, one seagul-network. Remember to only train duck-network on ducks, and seagul-network on seaguls.

Then you have to make a new model which is based on both. Call it seagul-duck-network. 
Now, do some model organ transplantation in seagul-duck-network.. Instead of feeding RoI's into the seagul-duck-mask-header, feed it into both the duck-mask-header and the seagul-mask-header, thus get two different header outoputs instead of one.

Note!
You dont want sharp edged basks but probabilities? You can modify the output of the mask head to give floats instead of bools, and use no thresholding. Thus you get a matrix of probabilities instead of a mask.

**Comment by zungam** [comment]
![image](https://user-images.githubusercontent.com/11159388/40540571-e68aa270-6018-11e8-9a38-5b4e67c2e72c.png)
Here is an image on what I mean

**Comment by enchantress1016** [comment]
@zungam Thank you so much, that's really inspiring answer! 
Let me make it clear, so the Mask RCNN is capable of output probability matrices for each RoI, and it'd do classification for each RoI after then. And if I need these probability matrices for RoIs, I need to modify the code to have float output instead of purely masks. 
I'll try start with one class problems to output probability matrices for RoIs, and later evolve to multiple classes problems.

**Comment by zungam** [comment]
Yes, I think so. Doing on a single class, the network needs to only be trained on this class. Output must be modified to floats instead of bools, with threshold=0, yes.

**Comment by zungam** [comment]
Btw, Mask-RCNN is made to be fast. If you dont need speed, you could also look into making an entire new architecture using only a end-to-end CNN (VGG16, Resnet101 as backbone) and just feed the entire image and get the entire image as output with instance masks. This must be with low resolution however to counteract how slow it is.

**Comment by enchantress1016** [comment]
@zungam Hi, I try to modify the code based on your idea and it works! Really appreciate for your help :)
The only thing I need to mention is that we don't necessarily need to adjust the value of DETECTION_MIN_CONFIDENCE. During my tests I found letting DETECTION_MIN_CONFIDENCE to be 0 may suggest you more RoIs, but you still can't get the probs for every the pixel. Letting it to be 0.9 gives more accurate results in my case.

For everyone else who is interested in having the probs of each RoIs, all you need to adjust are  detect_and_color_splash in your own config, and detect, unmold_detections functions in model file and unmold_mask in utils file.

**Comment by zxczhkzxc** [comment]
thank you for you problem,i want to ask something about mask-rcnn,as we all know mask rcnn is object-detection and mask are  parallel run,if i have different color of same class and i want to divide them,as i try use the mask-rcnn is does't work well how can i solve it?Could you elaborate on explaining how I could achieve this idea? Your help is really appreciated.@enchantress1016@zungam

**Comment by zungam** [comment]
@enchantress1016 no problem! :) Glad it worked. Did you manage to get the prob of each pixel inside each detection at least? I was actually aware that my approach wouldnt give a prob of the entire image.  For some reason I dont understand, I was thinking that you wanted the pixel for each RoI. Weird! Hmm.
Setting  DETECTION_MIN_CONFIDENCE to 0.9 might help as you fill out most of the image as you state.

@zxczhkzxc you are saying you trained a network on a class of some color? And now you want it to detect something with a new color? This can't be done unless you fine-tune it with new training data with the new updated color. If you absolutly dont have such training data, then you should modify the input images with photoshop or some script to match the new color (the surroundings must still be the old color).

**Comment by zxczhkzxc** [comment]
@zungam thanks for your reply,yeah my mind is to detect different color of one class and want to divide them,so i should have different color of this class as training data? when i use the fast-rcnn,it can detect only the one color as my training data(i don't have other color  training data ),that's surprised me,and i don't why?

**Comment by zungam** [comment]
@zxczhkzxc the network can only see the things you have trained it on, so if it hasnt seen any examples of the new color version, it will have a hard time knowing what it is

**Comment by zxczhkzxc** [comment]
got it ,thank you

**Comment by Xiaoyang-Rebecca** [comment]
mrcnn_checkpoint = model.run_graph([image], [
 ("detections", model.keras_model.get_layer("mrcnn_detection").output),
... ])
mrcnn_checkpoint['detection']     # is the matrix you need

**Comment by enchantress1016** [comment]
@Xiaoyang-Rebecca Hey, thank you for your reply. However, I tested the part you mentioned but it didn't seem right. I believe the output of mrcnn_checkpoint['detections'] are the probabilities of RoI but not pixel by pixel probabilities :)

**Comment by Xiaoyang-Rebecca** [comment]
Are you looking for the one for whole image or rois?

**Comment by gwangjinjeong** [comment]
@enchantress1016 Hi I read all comments but I can not convince my result that convert bool into the float. 
I modified the utils.py.
--Before
```python
    threshold = 0.5
    y1, x1, y2, x2 = bbox
    mask = resize(mask, (y2 - y1, x2 - x1))
    mask = np.where(mask >= threshold, 1, 0).astype(np.bool)

    # Put the mask in the right location.
    full_mask = np.zeros(image_shape[:2], dtype=np.bool)
    full_mask[y1:y2, x1:x2] = mask
    return full_mask
```
--After
```python
    y1, x1, y2, x2 = bbox
    mask = resize(mask, (y2 - y1, x2 - x1))
    full_mask = np.zeros(image_shape[:2], dtype=np.float32)
    full_mask[y1:y2, x1:x2] = mask
    return full_mask
```
But Could you give me the advice that is it right?
and Do you know how to get the probability in the background?

**Comment by kipronokoech** [comment]
@enchantress1016 , did you managed to get the answer to this issue?

I also need per-pixel probabilities for all classes for the whole image. The expected array should be of shape like (image width, image height, no_classes).

**Comment by colt18** [comment]
@gwangjinjeong Your approach is correct. This will return probability mask instead of boolean mask for the mentioned instances.

@kipronokoech You can achieve this by taking the per pixel probabilites of each instance and embedding them on to a final array. Say you you have `n classes`. For a single image file `img`, create `arr=np.zeros(w,h,n)`. Get your results by `results=model.detect([img], verbose=1)`. Then maintain a loop to;
```
    for each class_id in results[0]['class_ids]:
        get the index of nonzero elements
        append these elements to corresponding position in arr
```
